File: C++ Space Shooter src/asteroid.cpp
#include "asteroid.h"
#include "utils.h"
#include "image.h"

#include "raymath.h"

SpaceShooter::Image* largeAsteroidImage = nullptr;
SpaceShooter::Image* smallAsteroidImage = nullptr;

Asteroid::Asteroid(AsteroidTypes type, Vector2 pos, float direction)
{
	this->type = type;

	this->pos = pos;
	rotation = GetRandomValue(0, 360);

	switch (type)
	{
	case AsteroidTypes::Large:
		damage = 15;
		health = 15;

		rotationSpeed = (float)GetRandomValue(-120, 120) / 100.0f;

		speed = (float)GetRandomValue(25, 35) / 5.0f;

		movement = GetAngledOffset(speed, direction);

		radius = largeAsteroidImage->GetSize().x / 2.0f;
		break;
	case AsteroidTypes::Small:
		health = 8;
		damage = 8;

		rotationSpeed = (float)GetRandomValue(-180, 180) / 100.0f;

		speed = (float)GetRandomValue(30, 40) / 5.0f;

		movement = GetAngledOffset(speed, direction);

		radius = smallAsteroidImage->GetSize().x / 2.0f;
		break;
	}
}

void Asteroid::Update(float delta)
{
	pos = Vector2Add(pos, Vector2Multiply(movement, { delta, delta }));

	rotation += rotationSpeed * delta;
}

void Asteroid::Draw()
{
	switch (type)
	{
	case AsteroidTypes::Large:
		largeAsteroidImage->Draw(pos, rotation);
		break;
	case AsteroidTypes::Small:
		smallAsteroidImage->Draw(pos, rotation);
		break;
	}
}

void LoadAsteroidTextures()
{
	largeAsteroidImage = new SpaceShooter::Image("resources/large asteroid.png", 5.0f);
	smallAsteroidImage = new SpaceShooter::Image("resources/small asteroid.png", 5.0f);
}

void CleanUpAsteroidTextures()
{
	delete largeAsteroidImage;
	delete smallAsteroidImage;
}

File: 2022-2023/Space Shooter/data/modules/asteroid.py
import random

import pygame

from data.modules.constants import AsteroidTypes
from data.modules.image import Image
from data.modules.utils import get_angled_offset


class Asteroid:
	def __init__(self, asteroid_type, pos, direction):
		self.pos = pygame.Vector2(pos)

		self.type = asteroid_type

		if asteroid_type == AsteroidTypes.Large:
			self.image = Image("large asteroid", angle=random.randint(0, 360), scale=5)
			self.movement = get_angled_offset(direction, random.randint(3, 7))
			self.spin = float(random.randint(-10, 10)) / 10.0

			self.health = 20

			self.radius = 100
		else:
			self.image = Image("medium asteroid", angle=random.randint(0, 360), scale=5)
			self.movement = get_angled_offset(direction, random.randint(4, 8))
			self.spin = float(random.randint(-10, 10)) / 5.0

			self.health = 10

			self.radius = 60

	def damage(self, damage=0):
		self.health -= damage
		if self.health <= 0:
			return True
		return False

	def update(self, delta):
		self.pos += self.movement * delta

		self.image.angle += self.spin * delta

	def draw(self, surface, scroll):
		self.image.draw(surface, self.pos - scroll)
